### DISCUSSION TITLE
Target when project AssemblyAttribute AssemblyInfo is generated

### DISCUSSION THREAD

**Discussion by mwpowellhtx** [dotnet/msbuild#7530] [Q&A] [answered]
When are the `AssemblyInfo.cs` attributes generated in response to `CSPROJ` level `<AssemblyAttribute/>` items? Anyone have any ideas?

Using `msbuild --targets:targets.txt` to try to glean that for the project(s) in question, I get the following output, just for instance. But I do not see anything that jumps out as an `AssemblyInfo` target, per se.

Also, I assume these are listed in order of execution.

Thank you...

```
ApplyPackageManagement
ApplyPackageManagementItems
BeforePackageManagement
AfterPackageManagement
_GetRequiredWorkloads
ResolvePackageDependenciesDesignTime
CollectSDKReferencesDesignTime
CollectResolvedSDKReferencesDesignTime
ResolveFrameworkReferences
ResolveProjectReferencesDesignTime2
ResolveFrameworkReferencesDesignTime
CollectPackageReferences
CollectSuggestedWorkloads
_CheckCompileDesignTimePrerequisite
CollectAnalyzersDesignTime
CollectResolvedCompilationReferencesDesignTime
CollectUpToDateCheckInputDesignTime
CollectUpToDateCheckOutputDesignTime
CollectUpToDateCheckBuiltDesignTime
CompileDesignTime
ComputeXamarinTargetFrameworks
GetTargetFrameworks
_ComputeTargetFrameworkItems
GetTargetFrameworksWithPlatformFromInnerBuilds
DispatchToInnerBuilds
Build
_SetBuildInnerTarget
Clean
Rebuild
Restore
GenerateRestoreGraphFile
CollectCentralPackageVersions
CollectPackageDownloads
CollectFrameworkReferences
_LoadRestoreGraphEntryPoints
_FilterRestoreGraphProjectInputItems
_GenerateRestoreGraph
_GenerateRestoreGraphProjectEntry
_GenerateRestoreSpecs
_GenerateDotnetCliToolReferenceSpecs
_GetProjectJsonPath
_GetRestoreProjectStyle
EnableIntermediateOutputPathMismatchWarning
_GetRestoreTargetFrameworksOutput
_GetRestoreTargetFrameworksAsItems
_GetRestoreSettings
_GetRestoreSettingsCurrentProject
_GetRestoreSettingsAllFrameworks
_GetRestoreSettingsPerFramework
_GenerateRestoreProjectSpec
_GenerateProjectRestoreGraph
_GenerateRestoreDependencies
_GenerateProjectRestoreGraphAllFrameworks
_GenerateProjectRestoreGraphCurrentProject
_GenerateProjectRestoreGraphPerFramework
_GenerateRestoreProjectPathItemsCurrentProject
_GenerateRestoreProjectPathItemsPerFramework
_GenerateRestoreProjectPathItems
_GenerateRestoreProjectPathItemsAllFrameworks
_GenerateRestoreProjectPathWalk
_GetAllRestoreProjectPathItems
_GetRestoreSettingsOverrides
_GetRestorePackagesPathOverride
_GetRestoreRepositoryPathOverride
_GetRestoreSourcesOverride
_GetRestoreFallbackFoldersOverride
_Is
```

**Comment by baronfel** [comment, marked as answer]
Check out the `GenerateAssemblyInfo` target in the [dotnet/sdk](https://github.com/dotnet/sdk/blob/df4211931dba88ed4d599e21ec03c193eb058ca2/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.GenerateAssemblyInfo.targets#L43-L69) repo. This particular functionality is provided by the SDK itself, not MSBuild.

**Reply by mwpowellhtx** [reply]
Huh that is interesting. I wonder why that does not appear in the listed targets? Assuming targets are reported as they are seen during the natural build pipeline.

**Reply by baronfel** [reply]
That's a good question indeed. Have you tried taking an [MSBuild Binary Log](https://aka.ms/binlog) and checked the targets run in the viewer?  That would be the most complete picture of what's going on.

**Reply by mwpowellhtx** [reply]
Indeed I have not; that's a bit more in depth of a report than I would like for the detail I am aiming for there, I think. But I may look into it after all. Thanks.

**Comment by rainersigwald** [comment]
> Also, I assume these are listed in order of execution.

They are _not_; in the general case there's no way to predict the execution order from an evaluated project. Looking at a build log shows you what projects did run and in what order, and is the only way to really know that.

**Reply by mwpowellhtx** [reply]
Okay, so fact is they are not. Let me ask y'all this, then; why would it not make sense to list them in order? Is that not the point, to be able to see and discern the target order, befores, afters, dependencies, etc? Thank you...

**Reply by rainersigwald** [reply]
Because of design decisions very early in MSBuild's lifetime, the order of targets is dynamically determined during the build. The detailed rules are described in the doc [Target Build Order](https://docs.microsoft.com/visualstudio/msbuild/target-build-order?view=vs-2022). Because of that dynamism, there's no way to sort the list of targets.

**Reply by mwpowellhtx** [reply]
Okay, but does that invalidate the question, i.e. to _LOG_ what was encountered **_at the time it was encountered_** (?). I cannot speak for others, but I think, objectively, that is what anyone has been asking about, for a long, long time.

**Reply by rainersigwald** [reply]
I don't understand what you're asking. The standard build log can log all targets that are executed, in order, at higher verbosities.

**Reply by mwpowellhtx** [reply]
The question is, **_the targets only_**; don't care about the verbosity, that is way TMI.

**Reply by baronfel** [reply]
The `--targets` switch only prints the 'entry target' list because it only prints what happens through project evaluation, not target execution. If you'd like to trace what targets are actually visited during building specific target(s), that's a valid request. Thankfully it's fairly simple to do so yourself, if you'd like to see how useful it might be to you.  It _should_ be a matter of [writing a small custom logger](https://docs.microsoft.com/en-us/visualstudio/msbuild/build-loggers?view=vs-2022) that logs based off of the [`TargetStarted`](https://github.com/dotnet/msbuild/blob/522dd3c4067098d8d56c82136f44beb479137ff6/src/Framework/IEventSource.cs#L121) event handler on an `IEventSource`, and then you're all set.
